PR: Improve usability of runtests locally and test output on the CIs

CAM-Gerlach opened the pull request:
<!--- Before submitting your pull request, --->
<!--- please complete as much as practicable of the following checklist: --->

### Pull Request Checklist

* [x] Read and followed this repo's [Contributing Guidelines](https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md)
* [x] Based this PR on the latest version of the correct branch (master or 3.x)
* [x] Followed [PEP 8](https://www.python.org/dev/peps/pep-0008/) code style
* [x] Ensured this pull request hasn't eliminated unrelated blank lines/spaces,
      modified the ``spyder/defaults`` directory, or added new icons/assets
* [x] Wrote at least one-line docstrings following PEP 257 for any new functions
* [x] Added at least one unit test covering the changes, if at all possible
* [x] Described the changes and the motivation for them below


### Developer Certificate of Origin Affirmation

By submitting this Pull Request or typing my name below, I affirm the
[Developer Certificate of Origin](https://developercertificate.org/)
with respect to both the content of the contribution itself and this post,
and understand I am releasing it under Spyder's MIT (Expat) license.


<!--- TYPE YOUR GITHUB USERNAME AFTER THE FOLLOWING STATEMENT ---!>
I certify the above statement is true and correct: C.A.M. Gerlach

<!--- Note that you (not Spyder) retain copyright ownership of your work, --->
<!--- and may license it to other parties under the terms of your choice. --->
<!--- Contact us before incorporating any content from external projects. --->


## Description of Changes

<!--- Explain what you've done and why. --->

This PR does a few things to improve usability of both runtests locally (that have been bothering me and perhaps others for quite some time), as well as the CI reports for users developing both Spyder 3 and Spyder 4, and does a little opportunistic cleanup/refacotring of code that hasn't been touched between ``3.x`` and ``master``.

Namely, it:

* Adds the ability to pass any arbitrary additional pytest options of the users choosing through to pytest with ``argparse``, as well as some user-friendly help text and error handling
* Makes the ``-x`` option the default only on the CIs, so local users can see full test output without hacking ``runtests.py`` and can still pass  ``-x`` or Ctrl-C the test cycle manually if they so desire
* Removes the remaining % file coverage by default, as it can easily be re-enabled locally in the case the user really desires it for every file, and is already displayed in far more accessible form via two different dedicated services when run the CIs and otherwise just adds extra clutter to the logs
* Does some minor opportunistic cleanup/refactoring of non-conflicting ``3.x``/``master`` code

It was made against ``3.x``since both branches are still being actively developed, and thus the local and CI test output and runtests API would be inconsistent between them and have extra clutter. There should only be one small conflict—the Spyder profiler coverage stuff was removed, which was already removed in ``master``).


<!--- Thanks for your help making Spyder better for everyone! --->

ccordoba12 commented:
> the Spyder profiler coverage stuff was removed, which was already removed in master

It was removed in master because all plugins are now inside `spyder.plugins`, but that's not the case for our 3.x branch because `spyder_profiler` is a package on its own and it lives out of the `spyder` one. So please restore coverage for it.

ccordoba12 commented on runtests.py:
```diff
@@ -5,49 +5,57 @@
-                   '--durations=10',
-                   '--cov=spyder',
-                   '--cov=spyder_profiler',
-                   '--cov-report=term-missing']
```
Please restore everything of what you removed here, except for `'--cov-report=term-missing'`. That's still in master:

https://github.com/spyder-ide/spyder/blob/master/runtests.py#L41-L42

CAM-Gerlach replied:
I'm aware; this PR when merged would remove it there as well, per one of its major goals:

> Removes the remaining % file coverage by default, as it can easily be re-enabled locally in the case the user really desires it for every file, and is already displayed in far more accessible form via two different dedicated services when run the CIs and otherwise just adds extra clutter to the logs

However, if you have some practical use case for still keeping them in every CI report that isn't equally or better served by Codecov, Coveralls and running it locally, then I'm all ears.

ccordoba12 replied:
I don't have a problem with coverage being reported, but I'm almost never run the entire test suite locally, just the things I'm testing at the moment.

But sure, we can remove this.

CAM-Gerlach replied:
Okay, sure.

ccordoba12 commented on runtests.py:
```diff
@@ -5,49 +5,57 @@
+        usage='python runtests.py [--run-slow] [-- pytest_args]')
+    test_parser.add_argument('--run-slow', action='store_true',
+                             default=run_slow,
+                             help='Run the slow (mainly GUI/functional) tests')
```
There's no need to have the `--run-slow` option defined here. It's handled in `conftest.py`, as the [Pytest docs](https://docs.pytest.org/en/latest/example/simple.html) recommend to do it:

https://github.com/spyder-ide/spyder/blob/3.x/conftest.py#L30

CAM-Gerlach replied:
The reason I kept it was so it'd be exposed to the user e.g. through ``runtests.py --help`` and have clearer error handling with ``argparse``, etc. However, I can ahead and eliminate it since it does make the code cleaner and it isn't that big of a deal.

ccordoba12 replied:
Yes, please remove it, but leave it in the `usage` text above. I think that's enough.

CAM-Gerlach replied:
Okay, sure; done.

ccordoba12 commented on runtests.py:
```diff
@@ -5,49 +5,57 @@
+    test_parser.add_argument('pytest_args', nargs='*',
+                             help="Args to pass to pytest")
+    test_args = test_parser.parse_args()
+    main(run_slow=test_args.run_slow, extra_args=test_args.pytest_args)
```
Given the above comment, this only needs to be

    main(extra_args=test_args.pytest_args)

CAM-Gerlach replied:
Yes

ccordoba12 replied:
Ok, please go ahead with this.

CAM-Gerlach replied:
Yup, done.

CAM-Gerlach commented on runtests.py:
```diff
@@ -5,49 +5,57 @@
+        pytest_args.append('-x')
+    elif extra_args:
+        pytest_args += extra_args
+    if run_slow and '--run-slow' not in pytest_args:
```
> pytest_args is a fixed list, so there's no need to have the second condition.

Actually it isn't, as you can see from the line just before this one—the user may have added it as one of the ``extra_args``, a very likely possibility after the change you request below.

Also, ``run_slow`` != ``RUN_CI`` on ``master`` as discussed above, so unless you want to restructure how that is handled, this would still have to be there.

ccordoba12 replied:
Ok, then this needs to be

    if RUN_CI and '--run-slow' not in pytest_args:

CAM-Gerlach replied:
Okay sure, I did  that and ended up just refactoring the section entirely to make the logic simpler in ``3.x``. When you merge, you'll need to change it to check for the extra constant you add, or just directly, as to whether it is running on a macOS CI.

ccordoba12 commented on runtests.py:
```diff
@@ -5,49 +5,57 @@
-if os.environ.get('CI', None) is not None or '--run-slow' in sys.argv:
-    run_slow = True
+RUN_CI = os.environ.get('CI', None) is not None
+run_slow = RUN_CI
```
This is very confusing and I don't like it. `run_slow` is not the same as `RUN_CI` (which is what you're equating here), so please find a better way to write this or just leave them separate:

    RUN_CI = os.environ.get('CI', None) is not None
    RUN_SLOW = False

CAM-Gerlach replied:
Okay, however you prefer. I just dropped ``run_slow`` entirely then since it doesn't actually do anything (different from ``RUN_CI``) in ``3.x``; I did it that way originally to make your life a little easier when merging it. When you merge, you'll need to do something about it so the ``run_slow`` tests skip on Darwin, but that's up to you ofc.

CAM-Gerlach commented:
> The difference between RUN_CI and run_slow is still confusing to me (as proposed in this PR).

Right, its understandably confusing since there is no difference in ``3.x``, while there is on ``master`` since the ``run_slow`` tests are skipped when running on the CI under Darwin which ``run_slow`` currently checks for. I just left it as is to make your life easier when merging, but on request I removed it and you'll need to check for it somewhere when you merge with ``master``.

ccordoba12 commented:
Thanks for the change @CAM-Gerlach! It looks much better now!

However, please notice that without the `--cov` options, Codecov and Coveralls don't report anything in the PR checks. Therefore, I think you need to put them back, but please do it only when `RUN_CI == True`

CAM-Gerlach commented:
> However, please notice that without the --cov options, Codecov and Coveralls don't report anything in the PR checks.

Thanks; good catch—I should have checked that. I added ``--cov=spyder`` and ``--cov=spyder-profiler`` to the ``if RUN_CI`` additional arguments list (without the ``--cov-report=term-missing`` that produced overly verbose text-mode reports, unless that's needed for Codecov/Coveralls—it was removed in master already so I assume not). I also added the ``--no-cov-on-fail`` option, which will avoid printing a full coverage report for failed runs (which can get particularly troublesome due to the 3x repeat runs on fail) and also wasting time and resources calculating coverage for a failed build.

ccordoba12 commented:
> I added --cov=spyder and --cov=spyder-profiler to the if RUN_CI additional arguments list

Yeah, that's fine :+1: 

> I also added the --no-cov-on-fail option

Even better, I didn't know about that option.
